cpp/merkletree/sparse_merkle_tree.cc

```diff
@@ -0,0 +1,216 @@
+#include "cpp/merkletree/sparse_merkle_tree.h"
+
+#include <algorithm>
+#include <stddef.h>
+#include <vector>
+
+#include "merkletree/merkle_tree_math.h"
+#include "util/util.h"
+
+using std::make_pair;
+using std::ostream;
+using std::ostringstream;
+using std::reverse;
+using std::string;
+using std::unique_ptr;
+using std::unordered_map;
+using std::vector;
+
+
+const vector<string>& GetNullHashes(const TreeHasher& hasher) {
+  static unique_ptr<const vector<string>> null_hashes;
+  if (!null_hashes.get()) {
+    vector<string> r{hasher.HashLeaf("")};
+    for (int i(1); i < hasher.DigestSize() * 8; ++i) {
+      r.emplace_back(hasher.HashChildren(r.back(), r.back()));
+    }
+    reverse(r.begin(), r.end());
+    null_hashes.reset(new vector<string>(r.begin(), r.end()));
```

Comment:
If `unique_ptr<>::reset` isn't atomic, this could be Bad (tm) if two threads were trying to race for the construction of the first `SparseMerkleTree`.

My recommendation is "don't do that". :wink: